compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -2334,3 +2316,186 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        self.pat_fields.get_key_value(&ident).map(|(&ident, _)| ident).ok_or(ident)
+    }
+
+    pub fn do_check(
+        mut self,
+        adt_def: ty::AdtDef<'tcx>,
+        variant: &'tcx ty::VariantDef,
+    ) -> (/* has_union */ bool, Vec<(&'tcx ty::FieldDef, Ident)>) {
+        match adt_def.adt_kind() {
+            ty::AdtKind::Enum | ty::AdtKind::Struct => self.check_struct_variant(variant, true),
+            ty::AdtKind::Union => self.check_union_variant(adt_def),
+        };
+        if !self.conflict_set.is_empty() || !self.absent_set.is_empty() {
+            self.tcx.dcx().emit_err(errors::UnionPatMultipleFields {
+                span: self.pat_span,
+                conflict_set: self.conflict_set,
+                absent_set: self.absent_set,
+            });
+        }
+        (self.has_union, self.unmentioned_fields)
+    }
+
+    fn check_struct_variant(
+        &mut self,
+        variant: &'tcx ty::VariantDef,
+        collect_unmentioned: bool,
+    ) -> bool {
+        let mut has_mentioned = false;
+        for (idx, field) in variant.fields.iter_enumerated() {
+            if field.is_unnamed() {
+                let nested_adt_def = field.nested_adt_def(self.tcx);
+                has_mentioned |= if nested_adt_def.is_union() {
+                    self.check_union_variant(nested_adt_def)
+                } else {
+                    self.check_struct_variant(
+                        nested_adt_def.non_enum_variant(),
+                        collect_unmentioned,
+                    )
+                };
+            } else if self.field_mentioned(field).is_ok() {
+                has_mentioned = true;
+            } else {
+                return self.collect_unmentioned(
+                    variant.fields.iter().skip(idx.index()).take(usize::MAX),
+                    collect_unmentioned,
+                ) || has_mentioned;
+            }
+        }
+        has_mentioned
+    }
+
+    fn check_union_variant(&mut self, adt_def: ty::AdtDef<'tcx>) -> bool {
+        self.has_union = true;
+        let variant = adt_def.non_enum_variant();
+        let mut mentioned = None;
+        for (idx, field) in variant.fields.iter_enumerated() {
+            let current_mentioned = if field.is_unnamed() {
+                let nested_adt_def = field.nested_adt_def(self.tcx);
+                if nested_adt_def.is_union() {
+                    self.check_union_variant(nested_adt_def)
+                        .then_some((idx, /* need_recheck */ false))
+                } else {
+                    // dummy check, does not collect the unmentioned fields.
+                    // In case of the multiple fields from the union are used,
+                    // we do not collect those fields to report unmentioned errors.
+                    self.check_struct_variant(nested_adt_def.non_enum_variant(), false)
+                        .then_some((idx, /* need_recheck */ true))
```

Comment:
Similarly, I'm not sure that it's that big of a deal if we report an error for unmentioned fields from two unnamed struct types of an unnamed union (instead of just one of them). It may not be worth the additional complexity this adds.